rules/ruleset.go

```diff
@@ -71,84 +69,115 @@ func newActiveRuleSet(
 	uniqueCutoverTimes := make(map[int64]struct{})
 	for _, mappingRule := range mappingRules {
 		for _, snapshot := range mappingRule.snapshots {
-			uniqueCutoverTimes[snapshot.cutoverNs] = struct{}{}
+			uniqueCutoverTimes[snapshot.cutoverNanos] = struct{}{}
 		}
 	}
 	for _, rollupRule := range rollupRules {
 		for _, snapshot := range rollupRule.snapshots {
-			uniqueCutoverTimes[snapshot.cutoverNs] = struct{}{}
+			uniqueCutoverTimes[snapshot.cutoverNanos] = struct{}{}
 		}
 	}
 
-	cutoverTimeAsc := make([]int64, 0, len(uniqueCutoverTimes))
+	cutoverTimesAsc := make([]int64, 0, len(uniqueCutoverTimes))
 	for t := range uniqueCutoverTimes {
-		cutoverTimeAsc = append(cutoverTimeAsc, t)
+		cutoverTimesAsc = append(cutoverTimesAsc, t)
 	}
-	sort.Sort(int64Asc(cutoverTimeAsc))
+	sort.Sort(int64Asc(cutoverTimesAsc))
 
 	return &activeRuleSet{
-		version:         version,
 		iterFn:          iterFn,
 		newIDFn:         newIDFn,
 		mappingRules:    mappingRules,
 		rollupRules:     rollupRules,
-		cutoverTimesAsc: cutoverTimeAsc,
+		cutoverTimesAsc: cutoverTimesAsc,
 	}
 }
 
-func (as *activeRuleSet) Match(id []byte, t time.Time) MatchResult {
-	timeNs := t.UnixNano()
-	mappingCutoverNs, mappingPolicies := as.matchMappings(id, timeNs)
-	rollupCutoverNs, rollupResults := as.matchRollups(id, timeNs)
+// NB(xichen): could make this more efficient by keeping track of matched rules
+// at previous iteration and incrementally update match results.
+func (as *activeRuleSet) MatchAll(id []byte, from time.Time, to time.Time) MatchResult {
```

Comment:
Yeah in most cases I'd like to use unix nanoseconds since that's more efficient than storing or passing `time.Time` around, but if there is time-related logic that can't be cheaply done with unix nanoseconds, especially when the API is on the critical path, I'll still use `time.Time` to avoid conversion between unix nanoseconds and `time.Time`. Also, if it's an external API that might be used from multiple repos as opposed to an internal API that's only used inside a package, I try to use `time.Time` since that might be more friendly to callers.